pkg/frontend/querymiddleware/prune_test.go

```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/prometheus/promql/promqltest"
+	"github.com/stretchr/testify/require"
+)
+
+func TestQueryPruning(t *testing.T) {
+	endTime := 100
+	seriesName := `test_float`
+	const data = `
+		load 1m
+			test_float{series="1"} 0+1x100
+			test_float{series="2"} 0+2x100
+			test_float{series="3"} 0+3x100
+			test_float{series="4"} 0+4x100
+			test_float{series="5"} 0+5x100
+	`
+	queryable := promqltest.LoadedStorage(t, data)
+
+	const step = 20 * time.Second
+
+	engine := newEngine()
+	pruningware := newPruneMiddleware(
+		log.NewNopLogger(),
+	)
+	downstream := &downstreamHandler{
+		engine:    engine,
+		queryable: queryable,
+	}
+
+	type template struct {
+		query   string
+		IsEmpty bool
+	}
+
+	templates := []template{
+		{`avg(rate(%s[1m])) < (-1 * +Inf)`, true},
+		{`avg(rate(%s[1m])) < (+1 * +Inf)`, false},
+		{`avg(rate(%s[1m])) < (-1 * -Inf)`, false},
+		{`avg(rate(%s[1m])) < (+1 * -Inf)`, true},
+		{`(-1 * -Inf) < avg(rate(%s[1m]))`, true},
+		{`((-1 * -Inf) < avg(rate(foo[2m]))) or avg(rate(%s[1m]))`, false},
+		{`((-1 * -Inf) < avg(rate(foo[2m]))) and avg(rate(%s[1m]))`, true},
+		{`((-1 * -Inf) < avg(rate(foo[2m]))) unless avg(rate(%s[1m]))`, true},
+		{`avg(rate(%s[1m])) unless ((-1 * -Inf) < avg(rate(foo[2m])))`, false},
+		{`(((-1 * -Inf) < avg(rate(foo[3m]))) unless avg(rate(foo[2m]))) or avg(rate(%s[1m]))`, true},
+		{`((((-1 * -Inf) < avg(rate(foo[4m]))) unless avg(rate(foo[3m]))) and avg(rate(foo[2m]))) or avg(rate(%s[1m]))`, true},
+	}
+	for _, template := range templates {
+		t.Run(template.query, func(t *testing.T) {
+			query := fmt.Sprintf(template.query, seriesName)
+			req := &PrometheusRangeQueryRequest{
+				path:      "/query_range",
+				start:     0,
+				end:       int64(endTime * 1000 * 60),
```

Comment:
[nit] This might be clearer as something like `(100 * time.Minute).Milliseconds()`